terraform-provider-constellation/internal/provider/cluster_resource.go

```diff
@@ -344,6 +345,37 @@ func (r *ClusterResource) Configure(_ context.Context, req resource.ConfigureReq
 	}
 }
 
+// ModifyPlan is called when the resource is planned for creation, updates, or deletion. This allows to set pre-apply
+// warnings and errors.
+func (r *ClusterResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	// Checks running on updates to the resource. (i.e. state and plan != nil)
+	if !req.Plan.Raw.IsNull() && !req.State.Raw.IsNull() {
+		// Read currentState supplied by Terraform runtime into the model
+		var currentState ClusterResourceModel
+		resp.Diagnostics.Append(req.State.Get(ctx, &currentState)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		// Read plannedState supplied by Terraform runtime into the model
+		var plannedState ClusterResourceModel
+		resp.Diagnostics.Append(req.Plan.Get(ctx, &plannedState)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		// Warn the user about possibly destructive changes in case microservice changes are to be applied.
+		if currentState.MicroserviceVersion.ValueString() != "" && // current version is set (i.e. no create operation)
```

Comment:
the microservice version is allowed to be empty (lockstep). what does the comment mean?

Reply:
True. This is now obsolete with the whole check only running on an upgrade, and wouldn't have worked with the lock-step behaviour you've mentioned anyways.

Reply:
Since the microservice version is optional and K8s and image are also lockstepped, we should to that.